Linkerd Destination crash if endpoint slices enabled

## Bug Report

### What is the issue?
Linkerd Destination crash if endpoint slices enabled (-enable-endpoint-slices=true)
### How can it be reproduced?
Run Linkerd Destination with -enable-endpoint-slices=true flag
### Logs, error output, etc
`
panic: assignment to entry in nil map
/linkerd-build/controller/api/destination/watcher/endpoints_watcher.go:616 +0x1b4
github.com/linkerd/linkerd2/controller/api/destination/watcher.(*EndpointsWatcher).Subscribe(0xc000342380, 0xc0013f51da, 0x7, 0xc0013f51c0, 0x19, 0x1f90, 0x0, 0x0, 0x1f50e80, 0xc0009ef830, ...)
`

#### `linkerd check` output

```text
your output here ...
```

### Environment

- Kubernetes Version: 1.20.1
- Cluster Environment: (GKE, AKS, kops, ...)
- Host OS: Debian 10
- Linkerd version: edge-21.1.1

### Possible solution

change 

```
func (pp *portPublisher) endpointSliceToAddresses(es *discovery.EndpointSlice) AddressSet {
	addresses := make(map[ID]Address)
	resolvedPort := pp.resolveESTargetPort(es.Ports)
	if resolvedPort == Port(0) {
		return AddressSet{}
	}
```

to 

```
func (pp *portPublisher) endpointSliceToAddresses(es *discovery.EndpointSlice) AddressSet {
	addressSet :=  AddressSet{
	   Addresses: make(map[ID]Address),
	}
	
	resolvedPort := pp.resolveESTargetPort(es.Ports)
	if resolvedPort == Port(0) {
		return addressSet
	}
```

### Additional context

--- a/controller/api/destination/watcher/endpoints_watcher.go
+++ b/controller/api/destination/watcher/endpoints_watcher.go
@@ -31,6 +31,8 @@ const (
 	targetServiceNamespace = "target_service_namespace"
 )
 
+const endpointTargetRefPod = "Pod"
+
 // TODO: prom metrics for all the queues/caches
 // https://github.com/linkerd/linkerd2/issues/2204
 
@@ -128,6 +130,8 @@ type (
 )
 
 var endpointsVecs = newEndpointsMetricsVecs()
+
+var undefinedEndpointPort = Port(0)
 
 // NewEndpointsWatcher creates an EndpointsWatcher and begins watching the
 // k8sAPI for pod, service, and endpoint changes. An EndpointsWatcher will
@@ -704,11 +708,17 @@ func metricLabels(resource interface{}) map[string]string {
 }
 
 func (pp *portPublisher) endpointSliceToAddresses(es *discovery.EndpointSlice) AddressSet {
-	addresses := make(map[ID]Address)
-	resolvedPort := pp.resolveESTargetPort(es.Ports)
-	if resolvedPort == Port(0) {
-		return AddressSet{}
+	addressSet := AddressSet{
+		TopologicalPref: pp.TopologyPref,
+		Labels:          metricLabels(es),
+		Addresses:       make(map[ID]Address),
 	}
+
+	resolvedPort := pp.resolveESTargetPort(es.Ports)
+	if resolvedPort == undefinedEndpointPort {
+		return addressSet
+	}
+
 	serviceID, err := getEndpointSliceServiceID(es)
 	if err != nil {
 		pp.log.Errorf("Could not fetch resource service name:%v", err)
@@ -739,13 +749,13 @@ func (pp *portPublisher) endpointSliceToAddresses(es *discovery.EndpointSlice) A
 					address.TopologyLabels[k] = v
 				}
 
-				addresses[id] = address
+				addressSet.Addresses[id] = address
 			}
 
 			continue
 		}
 
-		if endpoint.TargetRef.Kind == "Pod" {
+		if endpoint.TargetRef.Kind == endpointTargetRefPod {
 			for _, IPAddr := range endpoint.Addresses {
 				address, id, err := pp.newPodRefAddress(resolvedPort, IPAddr, endpoint.TargetRef.Name, endpoint.TargetRef.Namespace)
 				if err != nil {
@@ -757,17 +767,13 @@ func (pp *portPublisher) endpointSliceToAddresses(es *discovery.EndpointSlice) A
 					address.TopologyLabels[k] = v
 				}
 
-				addresses[id] = address
+				addressSet.Addresses[id] = address
 			}
 		}
 
 	}
 
-	return AddressSet{
-		Addresses:       addresses,
-		Labels:          metricLabels(es),
-		TopologicalPref: pp.TopologyPref,
-	}
+	return addressSet
 }
 
 func (pp *portPublisher) endpointsToAddresses(endpoints *corev1.Endpoints) AddressSet {
@@ -793,7 +799,7 @@ func (pp *portPublisher) endpointsToAddresses(endpoints *corev1.Endpoints) Addre
 				continue
 			}
 
-			if endpoint.TargetRef.Kind == "Pod" {
+			if endpoint.TargetRef.Kind == endpointTargetRefPod {
 				address, id, err := pp.newPodRefAddress(resolvedPort, endpoint.IP, endpoint.TargetRef.Name, endpoint.TargetRef.Namespace)
 				if err != nil {
 					pp.log.Errorf("Unable to create new address:%v", err)
@@ -851,7 +857,7 @@ func (pp *portPublisher) newPodRefAddress(endpointPort Port, endpointIP, podName
 
 func (pp *portPublisher) resolveESTargetPort(slicePorts []discovery.EndpointPort) Port {
 	if slicePorts == nil {
-		return Port(0)
+		return undefinedEndpointPort
 	}
 
 	switch pp.targetPort.Type {
@@ -864,7 +870,7 @@ func (pp *portPublisher) resolveESTargetPort(slicePorts []discovery.EndpointPort
 			}
 		}
 	}
-	return Port(0)
+	return undefinedEndpointPort
 }
 
 func (pp *portPublisher) resolveTargetPort(subset corev1.EndpointSubset) Port {
@@ -878,7 +884,7 @@ func (pp *portPublisher) resolveTargetPort(subset corev1.EndpointSubset) Port {
 			}
 		}
 	}
-	return Port(0)
+	return undefinedEndpointPort
 }
 
 func (pp *portPublisher) updatePort(targetPort namedPort) {

